src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Healthchecks for sources
+use anyhow::Context;
+use chrono::NaiveDateTime;
+use std::fmt::Display;
+use std::sync::Arc;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+use tokio::sync::Mutex;
+use tracing::trace;
+
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::{Listen, ListenEvent, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_repr::{Datum, GlobalId, Row, Timestamp};
+
+use crate::source::{CollectionMetadata, NowFn};
+use crate::types::sources::SourceData;
+
+/// The Healthchecker is responsible for tracking the current state
+/// of a Timely worker for a source, as well as updating the relevant
+/// state collection based on it.
+pub struct Healthchecker {
+    /// Name of the source (e.g. kafka-s1)
+    source_name: String,
+    /// Name of the upstream resource, if any (e.g. Kafka topic, file path)
+    upstream_name: Option<String>,
+    /// Internal ID of the source (e.g. s1)
+    source_id: String,
```

Comment:
Mostly because all the times this is going to be used later (building Rows and matching against Rows) it is going to compare against the string serialization of it, so I wanted to pay the conversion cost once. I can switch it to `GlobalId` if you think it's better, though!

Reply:
eh, no strong opinion. I just like strongly typed things and we could keep that "information" as long as possible.